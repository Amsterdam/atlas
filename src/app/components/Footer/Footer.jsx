import React from 'react';
import Link from 'redux-first-router-link';
import { routing } from '../../routes';

const FooterComponent = () => (
  <footer className="c-dashboard__footer o-max-width">
    <div className="c-dashboard__footer-content u-row">
      <div className="u-col-sm--8">
        <div className="u-margin__left--2 c-dashboard__proclaimer">
          <h2 className="c-dashboard__proclaimer-header">Proclaimer</h2>
          <p className="c-dashboard__proclaimer-paragraph">Gemeente Amsterdam
            biedt met City Data een voorziening waar data voor iedereen
            beschikbaar zijn en hergebruikt mogen worden. De gegevens
            worden aangeboden door gemeentelijke en externe partijen. De
            eigenaren van deze gegevens zijn primair verantwoordelijk
            voor de kwaliteit van hun gepubliceerde data. Opmerkingen en
            aanvullingen op de gegevens kunt u <a
              className="c-link--inverse"
              href="mailto:terugmelding.basisinformatie@amsterdam.nl"
            >melden</a>.
          </p>
          <p className="c-dashboard__proclaimer-paragraph">
            <Link
              href={{ type: routing.proclaimer.type }}
              to={{ type: routing.proclaimer.type }}
              className="o-btn o-btn--link--inverse"
            >Lees de hele proclaimer</Link>
          </p>
        </div>
      </div>
      <div className="u-col-sm--3 u-col-sm--offset-1">
        <div className="c-dashboard__social">
          <h2 className="c-dashboard__social-header">Volg de gemeente</h2>
          <div className="c-dashboard__social-link">
            <a
              className="o-btn--link--inverse"
              href="https://twitter.com/AmsterdamNL"
              rel="external noopener noreferrer"
              target="_blank"
            >Twitter</a>
          </div>
          <div className="c-dashboard__social-link">
            <a
              className="o-btn o-btn--link--inverse"
              href="https://www.facebook.com/gemeenteamsterdam"
              rel="external noopener noreferrer"
              target="_blank"
            >Facebook</a>
          </div>
          <div className="c-dashboard__social-link">
            <a
              className="o-btn o-btn--link--inverse"
              href="https://www.linkedin.com/company/gemeente-amsterdam"
              rel="external noopener noreferrer"
              target="_blank"
            >LinkedIn</a>
          </div>
          <div className="c-dashboard__social-link">
            <a
              className="o-btn o-btn--link--inverse"
              href="https://github.com/Amsterdam"
              rel="external noopener noreferrer"
              target="_blank"
            >GitHub</a>
          </div>
        </div>
      </div>
      <div className="u-clearfix" />
    </div>
  </footer>
);

export default FooterComponent;
