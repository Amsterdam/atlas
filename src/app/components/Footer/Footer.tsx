import {
  Column,
  CompactThemeProvider,
  Footer as FooterComponent,
  FooterBottom,
  FooterSection,
  FooterTop,
  Link,
  List,
  ListItem,
  Paragraph,
  Row,
  themeSpacing,
} from '@datapunt/asc-ui'
import React from 'react'
import styled from 'styled-components'
import { FOOTER_LINKS } from '../../../shared/config/config'
import { openFeedbackForm } from '../Modal/FeedbackModal'
import FooterLinks, { FooterLink } from './FooterLinks'

const StyledLink = styled(Link)`
  margin-bottom: ${themeSpacing(3)};
`

const StyledParagraph = styled(Paragraph)`
  margin-bottom: ${themeSpacing(5)};
`

const helpLinks: FooterLink[] = [
  ...FOOTER_LINKS.HELP,
  {
    order: 3,
    title: 'Feedback geven',
    id: 'feedback_3',
    onClick: openFeedbackForm,
  },
].sort((a, b) => (a.order > b.order ? 1 : -1))

const Footer: React.FC = () => (
  <CompactThemeProvider>
    <FooterComponent>
      <FooterTop>
        <Row>
          <Column wrap span={{ small: 1, medium: 2, big: 2, large: 4, xLarge: 4 }}>
            <FooterSection title="Colofon">
              {FOOTER_LINKS && <FooterLinks links={FOOTER_LINKS.COLOFON} />}
            </FooterSection>
          </Column>
          <Column wrap span={{ small: 1, medium: 2, big: 2, large: 4, xLarge: 4 }}>
            <FooterSection title="Volg de gemeente">
              {FOOTER_LINKS && <FooterLinks links={FOOTER_LINKS.SOCIAL} />}
            </FooterSection>
          </Column>
          <Column wrap span={{ small: 1, medium: 2, big: 2, large: 4, xLarge: 4 }}>
            <FooterSection title="Vragen">
              <StyledParagraph>
                Heeft u een vraag en kunt u het antwoord niet vinden op deze website? Of heeft u
                bevindingen? Neem dan contact met ons op.
              </StyledParagraph>
              {FOOTER_LINKS && <FooterLinks links={helpLinks} />}
            </FooterSection>
          </Column>
        </Row>
      </FooterTop>
      <FooterBottom>
        <List>
          <ListItem>
            <StyledLink variant="with-chevron" {...FOOTER_LINKS.PRIVACY}>
              {FOOTER_LINKS.PRIVACY.title}
            </StyledLink>
          </ListItem>
        </List>
      </FooterBottom>
    </FooterComponent>
  </CompactThemeProvider>
)

export default Footer
