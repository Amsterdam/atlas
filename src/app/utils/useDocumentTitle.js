import React, { useEffect } from 'react'
import { useSelector } from 'react-redux'
import { routing } from '../routes'
import { getLocationType } from '../../store/redux-first-router/selectors'

const TITLE = 'Data en informatie - Amsterdam'

function useDocumentTitle() {
  const locationType = useSelector(getLocationType)
  const { title: storeTitle } = Object.values(routing).find(value => value.type === locationType)

  const [documentTitle, setTitle] = React.useState(`${storeTitle} - ${TITLE}`)

  function setDocumentTitle(pageTitle, documentTitleData = []) {
    const newTitle = [pageTitle || storeTitle, ...documentTitleData, TITLE].join(' - ')
    setTitle(newTitle)
    document.title = newTitle

    return newTitle
  }

  useEffect(() => {
    setDocumentTitle(storeTitle)
  }, [setDocumentTitle, storeTitle])

  return {
    documentTitle,
    setDocumentTitle,
  }
}

export default useDocumentTitle
