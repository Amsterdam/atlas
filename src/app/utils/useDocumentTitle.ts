import { useState } from 'react'
import { useSelector } from 'react-redux'
import { routing } from '../routes'
import { getLocationType } from '../../store/redux-first-router/selectors'

const TITLE = 'Data en informatie - Amsterdam'

function useDocumentTitle() {
  const locationType = useSelector(getLocationType)
  const route = Object.values(routing).find((value) => value.type === locationType)
  const routeTitle = route?.title ?? null
  const [documentTitle, setTitle] = useState(routeTitle ? `${routeTitle} - ${TITLE}` : TITLE)

  function setDocumentTitle(pageTitle: string | boolean, documentTitleData: string[] = []) {
    const newTitle = [pageTitle || routeTitle, ...documentTitleData, TITLE].join(' - ')
    setTitle(newTitle)
    document.title = newTitle

    return newTitle
  }

  return {
    documentTitle,
    setDocumentTitle,
  }
}

export default useDocumentTitle
