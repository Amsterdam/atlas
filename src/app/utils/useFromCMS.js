import React from 'react'
import { getByUrl } from '../../shared/services/api/api'
import cmsNormalizer from '../../shared/services/cms/cms-normalizer'
import { routing } from '../routes'

function useFromCMS(config, id = false, normalize = true) {
  const [results, setResults] = React.useState(null)
  const [loading, setLoading] = React.useState(true)

  const fetchData = async endpoint => {
    setLoading(true)
    try {
      if (!endpoint) {
        // eslint-disable-next-line no-param-reassign
        endpoint = id ? config.endpoint(id) : config.endpoint()
      }

      const { fields } = config
      const data = await getByUrl(endpoint)

      let result = data

      console.log(result)

      if (normalize) {
        result = await cmsNormalizer(config.type, data, fields)
      }
      setResults(result)
    } catch (e) {
      // window.location.replace(routing.niet_gevonden.path)
    }

    setLoading(false)
    return results
  }

  return {
    loading,
    fetchData,
    results,
  }
}

export default useFromCMS
