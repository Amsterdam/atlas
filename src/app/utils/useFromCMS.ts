import React from 'react'
import { getByUrl } from '../../shared/services/api/api'
import cmsJsonApiNormalizer from '../../shared/services/cms/cms-json-api-normalizer'
import useNormalizedCMSResults from '../../normalizations/cms/useNormalizedCMSResults'

export type CMSConfig = {
  endpoint: Function
  fields?: Array<string>
}

export type CMSResults = {
  loading: boolean
  fetchData: Function
  results: Array<CMSResultItem>
  error: boolean
}

// More fields should be added to this type when other CMS content pages are migrated to TypeScript
export type CMSResultItem = {
  id: string
  type: string
  specialType?: string
  title: string
  shortTitle?: string
  teaser: string
  teaserImage?: string
  coverImage?: string
}

function useFromCMS(config: CMSConfig, id = false, normalizeFromJSONApi = true): CMSResults {
  const [results, setResults] = React.useState([])
  const [loading, setLoading] = React.useState(true)
  const [error, setError] = React.useState(false)

  const fetchData = async (endpoint: string): Promise<Array<Object>> => {
    setLoading(true)
    try {
      if (!endpoint) {
        // eslint-disable-next-line no-param-reassign
        endpoint = id ? config.endpoint(id) : config.endpoint()
      }

      const { fields } = config
      const data = await getByUrl(endpoint)

      let result = data
      if (normalizeFromJSONApi) {
        result = await cmsJsonApiNormalizer(data, fields)
      }

      result = useNormalizedCMSResults(result)
      setResults(result)
    } catch (e) {
      setError(true)
    }

    setLoading(false)
    return results
  }

  return {
    loading,
    fetchData,
    results,
    error,
  }
}

export default useFromCMS
