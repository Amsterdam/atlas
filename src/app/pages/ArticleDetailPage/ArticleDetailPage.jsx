import styled from '@datapunt/asc-core'
import {
  Article,
  EditorialBody,
  EditorialContent,
  EditorialHeader,
  EditorialMetaList,
  EditorialSidebar,
  Column,
  CustomHTMLBlock,
  Heading,
  List,
  ListItem,
  Link,
  Typography,
  Row,
  themeColor,
  Paragraph,
} from '@datapunt/asc-ui'
import React from 'react'
import { connect } from 'react-redux'
import download from 'downloadjs'
import SHARED_CONFIG from '../../../shared/services/shared-config/shared-config'
import { getLocationPayload } from '../../../store/redux-first-router/selectors'
import useFromCMS from '../../utils/useFromCMS'
import './ArticleDetailPage.scss'
import EditorialPage from '../../components/EditorialPage/EditorialPage'
import { toArticleDetail } from '../../../store/redux-first-router/actions'
import ContentContainer from '../../components/ContentContainer/ContentContainer'
import cmsConfig from '../../../shared/services/cms/cms.config'
import normalizeDownloadsObject from '../../../normalizations/cms/normalizeDownloadFiles'

const ListItemContent = styled.div`
  display: flex;
  flex-direction: column;

  & > * {
    font-weight: 700;
    color: inherit;
  }
`

/* istanbul ignore next */
const ArticleDetailPage = ({ id }) => {
  const { fetchData, results, loading } = useFromCMS(cmsConfig.ARTICLE, id)

  React.useEffect(() => {
    fetchData()
  }, [id])

  const {
    title,
    date,
    localeDate,
    body,
    coverImageUrl,
    field_downloads: downloads,
    field_links: links,
    field_byline: byline,
    field_slug: slug,
    field_intro: intro,
  } = results || {}

  const documentTitle = title && `Artikel: ${title}`
  const linkAction = slug && toArticleDetail(id, slug)

  const normalizedDownloads = normalizeDownloadsObject(downloads)

  const DownloadLink = styled(Link).attrs({
    type: 'button',
  })`
    text-align: left;
    background-color: ${themeColor(
      'tint',
      'level1',
    )}; // Buttons are grey by default on Safari and Firefox

    small {
      text-transform: uppercase;
      color: ${themeColor('tint', 'level6')};
    }
  `

  return (
    <EditorialPage {...{ documentTitle, loading, linkAction }} description={intro}>
      {!loading && (
        <div className="article">
          <Row className="article__row">
            <ContentContainer>
              <Article
                {...(coverImageUrl
                  ? {
                      image: typeof coverImageUrl === 'string' ? coverImageUrl : undefined,
                    }
                  : {})}
              >
                <Row className="article__row">
                  <EditorialContent>
                    <Column
                      wrap
                      span={{ small: 1, medium: 2, big: 5, large: 11, xLarge: 11 }}
                      push={{ small: 0, medium: 0, big: 1, large: 1, xLarge: 1 }}
                    >
                      <Column span={{ small: 1, medium: 2, big: 4, large: 7, xLarge: 7 }}>
                        <EditorialBody>
                          <EditorialHeader title={title}>
                            <EditorialMetaList
                              dateTime={date}
                              dateFormatted={localeDate}
                              fields={byline && [{ id: 1, label: byline }]}
                            />
                          </EditorialHeader>
                          <Paragraph strong>{intro}</Paragraph>
                          <CustomHTMLBlock body={body} />
                        </EditorialBody>
                      </Column>
                      <Column
                        span={{ small: 1, medium: 2, big: 2, large: 3, xLarge: 3 }}
                        push={{ small: 0, medium: 0, big: 0, large: 1, xLarge: 1 }}
                      >
                        <EditorialSidebar>
                          {normalizedDownloads && normalizedDownloads.length ? (
                            <>
                              <Heading as="h2">Downloads</Heading>
                              <List>
                                {normalizedDownloads.map(({ fileName, key, type, size, url }) => (
                                  <ListItem key={key}>
                                    <DownloadLink
                                      $as="button"
                                      onClick={() => {
                                        download(`${SHARED_CONFIG.CMS_ROOT}${url}`)
                                      }}
                                      variant="with-chevron"
                                    >
                                      <ListItemContent>
                                        <span>{fileName}</span>
                                        <Typography as="small">({`${type} ${size}`})</Typography>
                                      </ListItemContent>
                                    </DownloadLink>
                                  </ListItem>
                                ))}
                              </List>
                            </>
                          ) : null}
                          {links && links.length ? (
                            <>
                              <Heading as="h2">Links</Heading>
                              <List>
                                {links.map(({ uri, title: linkTitle }) => (
                                  <ListItem key={uri}>
                                    <Link variant="with-chevron" href={`${uri}`}>
                                      {linkTitle}
                                    </Link>
                                  </ListItem>
                                ))}
                              </List>
                            </>
                          ) : null}
                        </EditorialSidebar>
                      </Column>
                    </Column>
                  </EditorialContent>
                </Row>
              </Article>
            </ContentContainer>
          </Row>
        </div>
      )}
    </EditorialPage>
  )
}

const mapStateToProps = state => {
  const { id } = getLocationPayload(state)
  return {
    id,
  }
}

export default connect(
  mapStateToProps,
  null,
)(ArticleDetailPage)
