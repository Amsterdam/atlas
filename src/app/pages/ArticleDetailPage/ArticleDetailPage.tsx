import {
  Accordion,
  AccordionWrapper,
  Article,
  breakpoint,
  Column,
  CustomHTMLBlock,
  EditorialBody,
  EditorialContent,
  EditorialMetaList,
  EditorialSidebar,
  Heading,
  Link,
  List,
  ListItem,
  Paragraph,
  Row,
  themeColor,
  themeSpacing,
  Typography,
} from '@amsterdam/asc-ui'
import { useMatomo } from '@datapunt/matomo-tracker-react'
import { Fragment, FunctionComponent, useEffect, useMemo } from 'react'
import { useParams } from 'react-router-dom'
import styled, { css } from 'styled-components'
import environment from '../../../environment'
import normalizeDownloadsObject from '../../../normalizations/cms/normalizeDownloadFiles'
import { EDITORIAL_FIELD_TYPE_VALUES } from '../../../normalizations/cms/normalizeCMSResults'
import cmsConfig from '../../../shared/config/cms.config'
import ContentContainer from '../../components/ContentContainer/ContentContainer'
import EditorialPage from '../../components/EditorialPage/EditorialPage'
import LoadingSpinner from '../../components/LoadingSpinner/LoadingSpinner'
import ShareBar from '../../components/ShareBar/ShareBar'
import { toArticleDetail } from '../../links'
import getImageFromCms from '../../utils/getImageFromCms'
import useDownload from '../../utils/useDownload'
import useFromCMS from '../../utils/useFromCMS'
import { DoubleNormalizedResults } from '../../../normalizations/cms/types'
import ErrorMessage from '../../components/ErrorMessage/ErrorMessage'

const ListItemContent = styled.div`
  display: flex;
  flex-direction: column;
  width: 100%;

  & > * {
    font-weight: 700;
    color: inherit;
  }
`

const DownloadLink = styled(Link).attrs({
  type: 'button',
})`
  text-align: left;
  background-color: ${themeColor(
    'tint',
    'level1',
  )}; // Buttons are grey by default on Safari and Firefox

  small {
    text-transform: uppercase;
    color: ${themeColor('tint', 'level6')};
  }

  &:disabled {
    cursor: default;
  }
`

const StyledHeading = styled(Heading)<{ isContentType: boolean }>`
  ${({ isContentType }) =>
    isContentType &&
    css`
      margin-bottom: ${themeSpacing(4)};
    `}
`

const StyledAccordionHeading = styled(Heading)`
  margin-top: ${themeSpacing(7)};
`

const StyledContentContainer = styled(ContentContainer)<{ hasImage: boolean }>`
  ${({ hasImage }) =>
    hasImage &&
    css`
      @media screen and ${breakpoint('max-width', 'tabletM')} {
        margin-top: 0;
      }
    `}
`

const StyledRow = styled(Row)`
  @media screen and ${breakpoint('max-width', 'tabletM')} {
    padding-left: 0;
    padding-right: 0;
  }
`

const EditorialBodyStyled = styled(EditorialBody)`
  width: 100%;
`

const StyledAccordion = styled(Accordion)`
  margin-top: ${themeSpacing(2)};
`

const StyledLink = styled(Link)`
  & > * {
    pointer-events: none;
  }
`

const FileInfo = styled(Typography)`
  display: inline-flex;
`

const StyledLoadingSpinner = styled(LoadingSpinner)`
  width: 15px;
  margin-left: ${themeSpacing(2)};
`

const ArticleDetailPage: FunctionComponent = () => {
  const { id } = useParams<{ id: string }>()
  const { fetchData, results, loading, error } = useFromCMS<DoubleNormalizedResults>(
    cmsConfig.ARTICLE,
    id,
  )
  const [downloadLoading, downloadFile] = useDownload()

  const result = results

  useEffect(() => {
    fetchData()
  }, [id])

  const image = useMemo(
    () => (result?.coverImage ? getImageFromCms(result?.coverImage, 1200, 600) : null),
    [result],
  )

  const { trackEvent } = useMatomo()

  const documentTitle = result?.title && `Artikel: ${result?.title}`
  const link = useMemo(() => (result?.slug ? toArticleDetail(result?.id, result?.slug) : null), [
    result,
  ])

  const normalizedDownloads = useMemo(
    () => result?.field_downloads && normalizeDownloadsObject(result?.field_downloads),
    [result],
  )

  const isContentType = result?.field_type === EDITORIAL_FIELD_TYPE_VALUES.CONTENT

  if (loading) {
    return <LoadingSpinner />
  }

  if (!result) {
    return (
      <ErrorMessage
        absolute
        message="Er is een fout opgetreden bij het laden van deze pagina."
        buttonLabel="Probeer opnieuw"
        buttonOnClick={fetchData}
      />
    )
  }

  const {
    title,
    localeDate,
    localeDateFormatted,
    body,
    coverImage,
    links,
    field_byline: byline,
    field_accordions: accordions,
    intro,
    field_language: lang,
  } = result

  return (
    <EditorialPage
      documentTitle={documentTitle}
      loading={loading}
      link={link}
      title={title}
      lang={lang}
      error={error}
      image={coverImage}
      description={intro}
    >
      {!loading && (
        <StyledRow>
          <StyledContentContainer hasImage={!!image}>
            {/*
            // @ts-ignore */}
            <Article image={image}>
              <Row>
                <EditorialContent>
                  <Column
                    wrap
                    span={{ small: 1, medium: 2, big: 5, large: 11, xLarge: 11 }}
                    push={{ small: 0, medium: 0, big: 1, large: 1, xLarge: 1 }}
                  >
                    <Column span={{ small: 1, medium: 2, big: 4, large: 7, xLarge: 7 }}>
                      <EditorialBodyStyled>
                        <StyledHeading forwardedAs="h1" isContentType={!isContentType}>
                          {title}
                        </StyledHeading>
                        {isContentType && (
                          <EditorialMetaList
                            // @ts-ignore
                            dateTime={localeDate}
                            dateFormatted={localeDateFormatted}
                            // @ts-ignore
                            fields={byline && [{ id: 1, label: byline }]}
                          />
                        )}
                        {intro && <Paragraph strong dangerouslySetInnerHTML={{ __html: intro }} />}
                        {typeof body === 'string' && (
                          <CustomHTMLBlock body={body.replace('http://', 'https://')} />
                        )}
                        {accordions?.length && (
                          <AccordionWrapper>
                            {accordions.map(
                              ({
                                field_accordion_title: accordionTitle,
                                field_accordion_content: accordionContent,
                                field_accordion_intro: accordionIntro,
                                field_accordion_label: accordionLabel,
                              }) => (
                                <Fragment key={`${accordionTitle}_${accordionLabel}`}>
                                  {accordionTitle && (
                                    <StyledAccordionHeading forwardedAs="h3">
                                      {accordionTitle}
                                    </StyledAccordionHeading>
                                  )}
                                  {accordionIntro?.processed && (
                                    <CustomHTMLBlock body={accordionIntro.processed} />
                                  )}
                                  {accordionLabel && accordionContent?.processed && (
                                    <StyledAccordion title={accordionLabel}>
                                      <CustomHTMLBlock body={accordionContent.processed} />
                                    </StyledAccordion>
                                  )}
                                </Fragment>
                              ),
                            )}
                          </AccordionWrapper>
                        )}
                      </EditorialBodyStyled>
                    </Column>
                    <Column
                      span={{ small: 1, medium: 2, big: 2, large: 3, xLarge: 3 }}
                      push={{ small: 0, medium: 0, big: 0, large: 1, xLarge: 1 }}
                    >
                      <EditorialSidebar>
                        {normalizedDownloads && normalizedDownloads.length ? (
                          <>
                            <Heading as="h2">Downloads</Heading>
                            <List>
                              {normalizedDownloads.map(({ fileName, key, type, size, url }) => (
                                <ListItem key={key}>
                                  <DownloadLink
                                    disabled={downloadLoading}
                                    forwardedAs="button"
                                    onClick={() => {
                                      trackEvent({
                                        category: 'Download',
                                        action: 'artikel-download',
                                        name: fileName,
                                      })
                                      downloadFile(`${environment.CMS_ROOT}${url}`)
                                    }}
                                    inList
                                  >
                                    <ListItemContent>
                                      <span>{fileName}</span>
                                      <FileInfo as="small">
                                        ({`${type} ${size}`})
                                        {downloadLoading && <StyledLoadingSpinner size={15} />}
                                      </FileInfo>
                                    </ListItemContent>
                                  </DownloadLink>
                                </ListItem>
                              ))}
                            </List>
                          </>
                        ) : null}
                        {links && links.length ? (
                          <>
                            <Heading as="h2">Links</Heading>
                            <List>
                              {links.map(({ uri, title: linkTitle }) => (
                                <ListItem key={uri}>
                                  <StyledLink inList href={`${uri}`}>
                                    <span>{linkTitle}</span>
                                  </StyledLink>
                                </ListItem>
                              ))}
                            </List>
                          </>
                        ) : null}
                      </EditorialSidebar>
                    </Column>
                  </Column>
                  <Column
                    span={{ small: 1, medium: 2, big: 4, large: 11, xLarge: 11 }}
                    push={{ small: 0, medium: 0, big: 1, large: 1, xLarge: 1 }}
                  >
                    <ShareBar topSpacing={6} />
                  </Column>
                </EditorialContent>
              </Row>
            </Article>
          </StyledContentContainer>
        </StyledRow>
      )}
    </EditorialPage>
  )
}

export default ArticleDetailPage
