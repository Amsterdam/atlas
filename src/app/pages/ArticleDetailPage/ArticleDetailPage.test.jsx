import React from 'react'
import { mount, shallow } from 'enzyme'
import configureMockStore from 'redux-mock-store'
import { ThemeProvider } from '@datapunt/asc-ui'
import { useMatomo } from '@datapunt/matomo-tracker-react'
import ArticleDetailPage from './ArticleDetailPage'
import useFromCMS from '../../utils/useFromCMS'
import linkAttributesFromAction from '../../../shared/services/link-attributes-from-action/linkAttributesFromAction'
import Footer from '../../components/Footer/Footer'
import useDocumentTitle from '../../utils/useDocumentTitle'

jest.mock('../../utils/useFromCMS')
jest.mock('../../../shared/services/link-attributes-from-action/linkAttributesFromAction')
jest.mock('../../components/Footer/Footer')
jest.mock('../../utils/useDocumentTitle')
jest.mock('@datapunt/matomo-tracker-react')

describe('ArticleDetailPage', () => {
  const id = 3
  const href = 'https://this.is/a-link/this-is-a-slug'

  const mockData = {
    fetchData: jest.fn(),
    results: {
      drupal_internal__nid: 100,
      title: 'This is a title',
      created: '2015-05-05',
      body: {
        value: 'body text',
      },
      field_file_size: 'file size',
      field_file_type: 'pdf',
      field_publication_source: 'source',
      field_intro: 'intro',
      field_slug: 'slug',
      included: [
        { attributes: { uri: { url: 'https://cover-link' } } },
        { attributes: { uri: { url: 'https://cover-link' } } },
        { attributes: { uri: { url: 'https://document-link' } } },
        { attributes: { uri: { url: 'https://document-link' } } },
      ],
    },
  }

  const fetchDataMock = jest.fn()

  let store
  beforeEach(() => {
    linkAttributesFromAction.mockImplementation(() => ({ href }))
    Footer.mockImplementation(() => <></>)
    useDocumentTitle.mockImplementation(() => ({
      setDocumentTitle: jest.fn(),
      href,
    }))
    useMatomo.mockImplementation(() => ({ trackPageView: jest.fn(), href }))

    store = configureMockStore()({ location: { payload: { id } }, ui: { isPrintMode: false } })
  })

  afterEach(() => {
    jest.resetAllMocks()
  })

  it('should render the spinner when the request is loading', () => {
    useFromCMS.mockImplementation(() => ({
      loading: true,
    }))

    const component = shallow(<ArticleDetailPage store={store} />)
      .dive()
      .dive()

    const editorialPage = component.find('EditorialPage').at(0)
    expect(editorialPage.props().loading).toBeTruthy()
  })

  it('should call the fetchData function when the component mounts', () => {
    useFromCMS.mockImplementation(() => ({
      loading: true,
      fetchData: fetchDataMock,
    }))

    const component = mount(
      <ThemeProvider>
        <ArticleDetailPage store={store} />
      </ThemeProvider>,
      { context: { store } },
    )

    expect(fetchDataMock).toHaveBeenCalled()
    expect(component.find('ArticleDetailPage').props().id).toBe(id)
  })

  it('should render the publication when there are results', () => {
    useFromCMS.mockImplementation(() => mockData)
    const component = shallow(<ArticleDetailPage store={store} />)
      .dive()
      .dive()

    expect(component).toMatchSnapshot()
  })
})
