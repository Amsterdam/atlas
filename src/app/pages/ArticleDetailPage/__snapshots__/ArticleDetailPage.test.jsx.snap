// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleDetailPage should render the publication when there are results 1`] = `
<EditorialPage
  documentTitle="Artikel: This is a title"
  linkAction={null}
  loading={false}
>
  <div
    className="article"
  >
    <Row
      className="article__row"
      debug={false}
      debugColor="rgba(51,102,153,0.75)"
      hasMargin={true}
    >
      <ContentContainer
        className="content-container"
      >
        <Article>
          <Row
            className="article__row"
            debug={false}
            debugColor="rgba(51,102,153,0.75)"
            hasMargin={true}
          >
            <EditorialContent>
              <Column
                debug={false}
                debugColor="rgba(51,102,153,0.5)"
                order={0}
                push={
                  Object {
                    "big": 1,
                    "large": 1,
                    "medium": 0,
                    "small": 0,
                    "xLarge": 1,
                  }
                }
                span={
                  Object {
                    "big": 5,
                    "large": 11,
                    "medium": 2,
                    "small": 1,
                    "xLarge": 11,
                  }
                }
                wrap={true}
              >
                <Column
                  debug={false}
                  debugColor="rgba(51,102,153,0.5)"
                  order={0}
                  push={0}
                  span={
                    Object {
                      "big": 4,
                      "large": 7,
                      "medium": 2,
                      "small": 1,
                      "xLarge": 7,
                    }
                  }
                  wrap={false}
                >
                  <EditorialBody>
                    <Styled(Heading)
                      $as="h1"
                      isContentType={true}
                    >
                      This is a title
                    </Styled(Heading)>
                    <Paragraph
                      strong={true}
                    />
                    <CustomHTMLBlock
                      body={
                        Object {
                          "value": "body text",
                        }
                      }
                    />
                  </EditorialBody>
                </Column>
                <Column
                  debug={false}
                  debugColor="rgba(51,102,153,0.5)"
                  order={0}
                  push={
                    Object {
                      "big": 0,
                      "large": 1,
                      "medium": 0,
                      "small": 0,
                      "xLarge": 1,
                    }
                  }
                  span={
                    Object {
                      "big": 2,
                      "large": 3,
                      "medium": 2,
                      "small": 1,
                      "xLarge": 3,
                    }
                  }
                  wrap={false}
                >
                  <EditorialContent />
                </Column>
              </Column>
              <Column
                debug={false}
                debugColor="rgba(51,102,153,0.5)"
                order={0}
                push={
                  Object {
                    "big": 1,
                    "large": 1,
                    "medium": 0,
                    "small": 0,
                    "xLarge": 1,
                  }
                }
                span={
                  Object {
                    "big": 4,
                    "large": 11,
                    "medium": 2,
                    "small": 1,
                    "xLarge": 11,
                  }
                }
                wrap={false}
              >
                <Connect(ShareBar)
                  topSpacing={6}
                />
              </Column>
            </EditorialContent>
          </Row>
        </Article>
      </ContentContainer>
    </Row>
  </div>
</EditorialPage>
`;
